File: Code-CSharp/Base/Fstream/FstreamEntry.cs
namespace CodeCsharp.Base.Fstream;

using System.Text;

public class FstreamEntry {
  public static void FstreamProgram() {
    //initialize string
    string str = "output";

    //create file
    //open output file stream
    try {
      FileStream outFile = File.Open("file.txt", FileMode.Create, FileAccess.Write);
      StreamWriter writer = new StreamWriter(outFile);
      //whether open operation succeeds
      bool isOpen = true;
      //write a char
      writer.Write('c');
      //write a string with line seperator
      writer.WriteLine(str);
      //close
      outFile.Close();
    } catch (IOException exception) {
      bool isOpen = false;
    }

    //open input file stream
    try {
      FileStream inFile = File.Open("file.txt", FileMode.Open, FileAccess.Read);
      StreamReader reader = new StreamReader(inFile);
      //get first char
      char c1 = (char)reader.Read();
      //peek next char
      char c2 = (char)reader.Peek();
      //intialize a buffer
      char[] cb = new char[100];
      //read several char into buffer
      reader.Read(cb, 0, str.Length);
      //create string from buffer
      str = new string(cb);
      //close
      inFile.Close();
    } catch (IOException exception) {
    }

    //open append file stream
    try {
      FileStream appFile = File.Open("file.txt", FileMode.Append, FileAccess.Write);
      StreamWriter writer = new StreamWriter(appFile);
      //append a char
      writer.Write('c');
      //close
      appFile.Close();
    } catch (IOException exception) {
    }

    //open binary output file stream
    try {
      FileStream binaryOutFile = new FileStream("file.txt", FileMode.Create, FileAccess.Write);
      BinaryWriter dataOutFile = new BinaryWriter(binaryOutFile);
      //write string data in binary
      dataOutFile.Write(Encoding.UTF8.GetBytes(str));
      //write int data in binary
      int intNum = 10;
      dataOutFile.Write(intNum);
      //write double data in binary
      double doubleNum = 3.14;
      dataOutFile.Write(doubleNum);
      //close
      binaryOutFile.Close();
    } catch (IOException exception) {
    }

    //open binary input file stream
    try {
      FileStream binaryInFile = new FileStream("file.txt", FileMode.Open, FileAccess.Read);
      BinaryReader dataInFile = new BinaryReader(binaryInFile);
      //read string data in binary
      byte[] bytes = new byte[Encoding.UTF8.GetByteCount(str)];
      bytes = dataInFile.ReadBytes(bytes.Length);
      str = Encoding.UTF8.GetString(bytes);
      //read int data in binary
      int intNum = dataInFile.ReadInt32();
      //read double data in binary
      double doubleNum = dataInFile.ReadDouble();
      //get byte length of binary file
      long length = binaryInFile.Length;
      //close
      binaryInFile.Close();
    } catch (IOException exception) {
    }

    //open random access file stream
    try {
      //set position
      FileStream randomAccessFile = new FileStream("file.txt", FileMode.Open, FileAccess.ReadWrite);
      //get a char from position
      randomAccessFile.Seek(3, SeekOrigin.Begin);
      char c3 = (char)randomAccessFile.ReadByte();
      //get position
      long pos = randomAccessFile.Position;
      //write a char from position
      randomAccessFile.Write(new[] { (byte)'d' });
      //close
      randomAccessFile.Close();
    } catch (IOException exception) {
    }
  }

  public static bool TestAll() {
    bool ret = true;
    string str = "output";
    try {
      FileStream outFile = File.Open("test.txt", FileMode.Create, FileAccess.Write);
      StreamWriter writer = new StreamWriter(outFile);
      bool isOpen = true;
      writer.Write('c');
      writer.WriteLine(str);
    } catch (IOException exception) {
      bool isOpen = false;
    }

    try {
      FileStream appFile = File.Open("test.txt", FileMode.Append, FileAccess.Write);
      StreamWriter writer = new StreamWriter(appFile);
      writer.Write('d');
    } catch (IOException exception) {
    }

    try {
      FileStream inFile = File.Open("test.txt", FileMode.Open, FileAccess.Read);
      StreamReader reader = new StreamReader(inFile);
      ret = ret && (char)reader.Read() == 'c';
      ret = ret && (char)reader.Peek() == 'o';
      char[] cb = new char[100];
      reader.Read(cb, 0, str.Length);
      string str2 = new string(cb);
      ret = ret && str.Equals(str2);
      ret = ret && (char)reader.Read() == '\n';
      ret = ret && (char)reader.Read() == 'd';
    } catch (IOException exception) {
    }

    try {
      FileStream binaryOutFile =
          new FileStream("binary-test.txt", FileMode.Create, FileAccess.Write);
      BinaryWriter dataOutFile = new BinaryWriter(binaryOutFile);
      dataOutFile.Write(Encoding.UTF8.GetBytes(str));
      int intNum = 10;
      dataOutFile.Write(intNum);
      double doubleNum = 3.14;
      dataOutFile.Write(doubleNum);
    } catch (IOException exception) {
    }

    try {
      FileStream binaryInFile = new FileStream("binary-test.txt", FileMode.Open, FileAccess.Read);
      BinaryReader dataInFile = new BinaryReader(binaryInFile);
      byte[] bytes = new byte[Encoding.UTF8.GetByteCount(str)];
      bytes = dataInFile.ReadBytes(bytes.Length);
      string str2 = Encoding.UTF8.GetString(bytes);
      ret = ret && str.Equals(str2);
      int intNum = dataInFile.ReadInt32();
      ret = ret && intNum == 10;
      double doubleNum = dataInFile.ReadDouble();
      ret = ret && doubleNum == 3.14;
      long length = binaryInFile.Length;
      ret = ret && length == 18;
      binaryInFile.Close();
    } catch (IOException exception) {
    }

    try {
      FileStream randomAccessFile = new FileStream("test.txt", FileMode.Open, FileAccess.ReadWrite);
      randomAccessFile.Seek(3, SeekOrigin.Begin);
      ret = ret && (char)randomAccessFile.ReadByte() == 't';
      ret = ret && randomAccessFile.Position == 4;
      randomAccessFile.Write(new[] { (byte)'e' });
      randomAccessFile.Seek(4, SeekOrigin.Begin);
      ret = ret && (char)randomAccessFile.ReadByte() == 'e';
    } catch (IOException exception) {
    }

    if (!ret)
      Console.WriteLine("Fstream Test Failed!");
    return ret;
  }
}

File: Code-Cpp/code_cpp/base/fstream/fstream_entry.cc
#include "fstream_entry.h"
#include <iostream>
void FstreamEntry::FstreamProgram() {
  //initialize string
  std::string str = "output";
  //create file
  //open output file stream
  std::ofstream out_file("file.txt", std::ios::out | std::ios::trunc);
  //whether open operation succeeds
  bool is_open = !out_file;
  //write a char
  out_file.put('c');
  //write a string with line seperator
  out_file << str << std::endl;
  //close
  out_file.close();
  //open input file stream
  std::ifstream in_file("file.txt", std::ios::in);
  //get first char
  char c1 = in_file.get();
  //peek next char
  char c2 = in_file.peek();
  //intialize a buffer
  char cb[100];
  //read several char into buffer
  in_file.get(cb, 6 + 1);
  //create string from buffer
  str = std::string(cb, 6 + 1);
  //close
  in_file.close();
  //open append file stream
  std::ofstream app_file("file.txt", std::ios::out | std::ios::app);
  //append a char
  app_file.put('c');
  //close
  app_file.close();
  //open binary output file stream
  std::ofstream binary_out_file("binaryfile.txt", std::ios::out | std::ios::trunc | std::ios::binary);
  //write string data in binary
  binary_out_file.write(str.data(), str.size());
  //write int data in binary
  int int_num = 10;
  binary_out_file.write(reinterpret_cast<const char *>(&int_num), sizeof(int_num));
  //write double data in binary
  double double_num = 3.14;
  binary_out_file.write(reinterpret_cast<const char *>(&double_num), sizeof(double_num));
  //close
  binary_out_file.close();
  //open binary input file stream
  std::ifstream binary_in_file("binaryfile.txt", std::ios::in | std::ios::binary);
  //read string data in binary
  binary_in_file.read(str.data(), str.size());
  //read int data in binary
  binary_in_file.read(reinterpret_cast<char *>(&int_num), sizeof(int_num));
  //read double data in binary
  binary_in_file.read(reinterpret_cast<char *>(&double_num), sizeof(double_num));
  //get byte length of binary file
  binary_in_file.seekg(0, std::ios::end);
  long length = binary_in_file.tellg();
  //close
  binary_in_file.close();
  //open random access file stream
  std::fstream random_access_file("file.txt", std::ios::in | std::ios::out);
  //set position
  random_access_file.seekg(3);
  //get a char from position
  char c3 = random_access_file.get();
  //get position
  long pos = random_access_file.tellg();
  //write a char from position
  random_access_file.put('d');
  //close
  random_access_file.close();
}
bool FstreamEntry::TestAll() {
  bool ret = true;
  using namespace std;
  string str = "output";
  ofstream out_file("test.txt", ios::out | ios::trunc);
  ret = out_file.is_open();
  out_file.put('c');
  out_file << str << endl;
  out_file.close();
  ofstream app_file("test.txt", ios::out | ios::app);
  app_file.put('d');
  app_file.close();
  ifstream in_file("test.txt", ios::in);
  ret = ret && in_file.get() == 'c';
  ret = ret && in_file.peek() == 'o';
  char cb[100];
  in_file.get(cb, 6 + 1);
  string str2 = string(cb, 6 + 1);
  ret = ret && str == str2.substr(0, 6);
  ret = ret && in_file.get() == '\n';
  ret = ret && in_file.get() == 'd';
  in_file.close();

  std::ofstream binary_out_file("binary-test.txt", std::ios::out | std::ios::trunc | std::ios::binary);
  binary_out_file.write(str.data(), str.size());
  int int_num = 10;
  binary_out_file.write(reinterpret_cast<const char *>(&int_num), sizeof(int_num));
  double double_num = 3.14;
  binary_out_file.write(reinterpret_cast<const char *>(&double_num), sizeof(double_num));
  binary_out_file.close();
  std::ifstream binary_in_file("binary-test.txt", std::ios::in | std::ios::binary);
  binary_in_file.read(str2.data(), str.size());
  ret = ret && str == str2.substr(0, 6);
  binary_in_file.read(reinterpret_cast<char *>(&int_num), sizeof(int_num));
  ret = ret && int_num == 10;
  binary_in_file.read(reinterpret_cast<char *>(&double_num), sizeof(double_num));
  ret = ret && double_num == 3.14;
  binary_in_file.seekg(0, std::ios::end);
  long length = binary_in_file.tellg();
  ret = ret && length == 18;
  binary_in_file.close();

  std::fstream random_access_file("test.txt", std::ios::in | std::ios::out);
  random_access_file.seekg(3);
  ret = ret && random_access_file.get()=='t';
  ret = ret && random_access_file.tellg() == 4;
  random_access_file.seekg(4);
  random_access_file.put('e');
  random_access_file.seekg(4);
  ret = ret && random_access_file.get()=='e';
  random_access_file.close();

  if (!ret)
    cout << "Fstream Test Failed!";
  return ret;
}
